src/balancerd/src/lib.rs

```diff
@@ -395,19 +418,78 @@ impl mz_ore::server::Server for PgwireBalancer {
     }
 }
 
-struct HttpBalancer {
-    _resolver: Arc<Resolver>,
+struct HttpsBalancer {
+    tls: Arc<Option<SslContext>>,
+    resolve_template: Arc<String>,
+    sni_extract: Arc<Regex>,
     // todo: metrics
 }
 
-impl mz_ore::server::Server for HttpBalancer {
-    const NAME: &'static str = "http_balancer";
+impl mz_ore::server::Server for HttpsBalancer {
+    const NAME: &'static str = "https_balancer";
+
+    fn handle_connection(&self, conn: TcpStream) -> mz_ore::server::ConnectionHandler {
+        let tls_context = Arc::clone(&self.tls);
+        let resolve_template = Arc::clone(&self.resolve_template);
+        let sni_extract = Arc::clone(&self.sni_extract);
+        Box::pin(async move {
+            let (mut client_stream, servername): (Box<dyn ClientStream>, Option<String>) =
+                match &*tls_context {
+                    Some(tls_context) => {
+                        let mut ssl_stream = SslStream::new(Ssl::new(tls_context)?, conn)?;
+                        if let Err(e) = Pin::new(&mut ssl_stream).accept().await {
+                            let _ = ssl_stream.get_mut().shutdown().await;
+                            return Err(e.into());
+                        }
+                        let servername :Option<String>=
+                            ssl_stream.ssl().servername(NameType::HOST_NAME).map(|sn| {
+                                debug!("full servername: {sn}");
+                                match sni_extract
+                                    .captures(sn)
+                                    .map(|caps| caps.get(1))
+                                    .flatten() {
+                                    Some(cap) => cap.as_str().to_string(),
+                                    None => {
+                                        debug!("sni extract regex didn't match or didn't have a capture group");
+                                        sn.to_string()
+                                    }
+                                }
+                            });
+                        debug!("servername: {servername:?}");
+                        (Box::new(ssl_stream), servername)
+                    }
+                    _ => (Box::new(conn), None),
+                };
+
+            let addr: Cow<String> = match servername {
+                Some(servername) => {
+                    let name = resolve_template.replace("{}", &servername);
```

Comment:
I don't love the string replacement, what do you think about turning `resolve_template` into a struct?, e.g.

```
struct HttpsResolveTemplate {
  host: String,
  name: String,
  port: String,
}

impl HttpsResolveTemplate {
  fn generate(&self, servername: &str) -> String {
    let HttpsResolveTemplate { host, name, port } = self;
    format!("{host}.{servername}.{name}:{port}")
  }
}
```

Reply:
Ahhhh okay that makes sense, yeah in this case I think the string replace works, thanks for clarifying!